services/semantic-memory/lib-dotnet/Storage/ContentStorage/AzureBlob.cs

```diff
@@ -0,0 +1,202 @@
+            new StorageSharedKeyCredential(accountName, accountKey)), containerName, logger)
+    {
+    }
+
+    public AzureBlob(BlobServiceClient client, string containerName = "skmemory", ILogger<AzureBlob>? logger = null)
+    {
+        if (string.IsNullOrEmpty(containerName))
+        {
+            throw new ContentStorageException("The container name is empty");
+        }
+
+        this._containerName = containerName;
+        this._containerClient = client.GetBlobContainerClient(containerName);
+
+        if (this._containerClient == null)
+        {
+            throw new ContentStorageException("Unable to instantiate Azure Blob container client");
+        }
+
+        this._log = logger ?? NullLogger<AzureBlob>.Instance;
+    }
+
+    /// <inherit />
+    public async Task CreateDirectoryAsync(string directoryName, CancellationToken cancellationToken = default)
+    {
+        this._log.LogTrace("Creating container '{0}' ...", this._containerName);
+
+        await this._containerClient
+            .CreateIfNotExistsAsync(PublicAccessType.None, cancellationToken: cancellationToken)
+            .ConfigureAwait(false);
+
+        this._log.LogTrace("Container '{0}' ready", this._containerName);
+    }
+
+    /// <inherit />
+    public Task WriteTextFileAsync(string directoryName, string fileName, string fileContent, CancellationToken cancellationToken = default)
+    {
+        return this.InternalWriteAsync(directoryName, fileName, fileContent, cancellationToken);
+    }
+
+    /// <inherit />
+    public Task<long> WriteStreamAsync(string directoryName, string fileName, Stream contentStream, CancellationToken cancellationToken = default)
+    {
+        return this.InternalWriteAsync(directoryName, fileName, contentStream, cancellationToken);
+    }
+
+    /// <inherit />
+    public async Task<BinaryData> ReadFileAsync(string directoryName, string fileName, CancellationToken cancellationToken = default)
+    {
+        var blobName = $"{directoryName}/{fileName}";
+        BlobClient blobClient = this.GetBlobClient(blobName);
+        Response<BlobDownloadResult>? content = await blobClient.DownloadContentAsync(cancellationToken).ConfigureAwait(false);
+
+        if (content == null || !content.HasValue)
+        {
+            this._log.LogError("Unable to download file {0}", blobName);
+            throw new ContentStorageException("Unable to fetch blob content");
+        }
+
+        return content.Value.Content;
+    }
+
+    private async Task<long> InternalWriteAsync(string directoryName, string fileName, object content, CancellationToken cancellationToken)
+    {
+        var blobName = $"{directoryName}/{fileName}";
+
+        BlobClient blobClient = this.GetBlobClient(blobName);
+
+        BlobUploadOptions options = new();
+        BlobLeaseClient? blobLeaseClient = null;
+        BlobLease? lease = null;
+        if (await blobClient.ExistsAsync(cancellationToken).ConfigureAwait(false))
```

Comment:
nit: It might be faster to skip the call to ExistsAsync and instead just try to get the leaseClient and lease. If it doesn't exist it will throw and the leaseClient/lease will remain null.

Reply:
interesting idea, that should reduce the number of requests, thanks. LeaseBlobAsync() throws exceptions also when the file is already leased, so the code will also have to check the status code (404 vs 409 probably). Keeping this open while I do some tests